crates/c-api/src/wasi.rs

```diff
@@ -0,0 +1,274 @@
+    };
+
+    let mut config = Box::from_raw(config);
+    config.builder = config.builder.stderr(file);
+    std::mem::forget(config);
+
+    true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_config_inherit_stderr(config: *mut wasi_config_t) {
+    let mut config = Box::from_raw(config);
+    config.builder = config.builder.inherit_stderr();
+    std::mem::forget(config);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_config_preopen_dir(
+    config: *mut wasi_config_t,
+    path: *const c_char,
+    guest_path: *const c_char,
+) -> bool {
+    let guest_path = match cstr_to_path(guest_path) {
+        Some(p) => p,
+        None => return false,
+    };
+
+    let dir = match cstr_to_path(path) {
+        Some(p) => match preopen_dir(p) {
+            Ok(d) => d,
+            Err(_) => return false,
+        },
+        None => return false,
+    };
+
+    let mut config = Box::from_raw(config);
+    config.builder = config.builder.preopened_dir(dir, guest_path);
+    std::mem::forget(config);
+
+    true
+}
+
+#[repr(C)]
+pub struct wasi_instance_t {
+    wasi: Wasi,
+    export_cache: HashMap<String, *mut wasm_extern_t>,
+}
+
+impl Drop for wasi_instance_t {
+    fn drop(&mut self) {
+        for v in self.export_cache.values() {
+            drop(unsafe { Box::from_raw(*v) });
+        }
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_instance_new(
+    store: *mut wasm_store_t,
+    config: *mut wasi_config_t,
+    trap: *mut *mut wasm_trap_t,
+) -> *mut wasi_instance_t {
+    let store = &(*store).store.borrow();
+    let config = Box::from_raw(config);
+
+    match config.builder.build() {
+        Ok(ctx) => Box::into_raw(Box::new(wasi_instance_t {
+            wasi: Wasi::new(store, ctx),
+            export_cache: HashMap::new(),
+        })),
+        Err(e) => {
+            (*trap) = Box::into_raw(Box::new(wasm_trap_t {
+                trap: HostRef::new(Trap::new(e.to_string())),
+            }));
+
+            std::ptr::null_mut()
+        }
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_instance_delete(instance: *mut wasi_instance_t) {
+    drop(Box::from_raw(instance));
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasi_instance_bind_import(
+    instance: *mut wasi_instance_t,
+    import: *const wasm_importtype_t,
+) -> *const wasm_extern_t {
+    // TODO: support previous versions?
+    if (*import).ty.module() != "wasi_snapshot_preview1" {
+        return std::ptr::null_mut();
+    }
+
+    // The import should be a function (WASI only exports functions)
+    let func_type = match (*import).ty.ty() {
+        ExternType::Func(f) => f,
+        _ => return std::ptr::null_mut(),
+    };
+
+    let name = (*import).ty.name();
+
+    match (*instance).wasi.get_export(name) {
+        Some(export) => {
+            if export.ty().params() != func_type.params() {
+                return std::ptr::null_mut();
+            }
+
+            if export.ty().results() != func_type.results() {
```

Comment:
Should we perhaps implement `PartialEq` for `FuncType` itself so you don't have to manually check the params/results?

Reply:
That's a good idea.  I'll add that and replace this with a type comparison.

Reply:
Fixed.